synapse/handlers/sync.py

```diff
@@ -376,7 +377,33 @@ def ephemeral_by_room(self, sync_result_builder, now_token, since_token=None):
                 room_id = event["room_id"]
                 # exclude room id, as above
                 event_copy = {k: v for (k, v) in iteritems(event) if k != "room_id"}
-                ephemeral_by_room.setdefault(room_id, []).append(event_copy)
+
+                # filter out receipts the user shouldn't see
+                content = event_copy.get("content", {})
```

Comment:
I've not really managed to grok what this lump of code is doing. If it's just filtering events, shouldn't that be handled in `receipt_source.get_new_events` ?

Reply:
Ideally yes, however there's a comment in a similar area around typing notifications that says the results from that are cached. Didn't want to risk all read receipts being dropped due to a caching issue.

Reply:
the comment says you shouldn't modify returned event objects, which is true... I don't think there's any reason you can't modify `get_new_events` to return a different set of events (modulo checking where else it is used).

This code feels super-complicated at the moment.